digester.go

```diff
@@ -0,0 +1,267 @@
+
+package parser
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	hash2 "hash"
+	"strings"
+	"sync"
+	"unicode"
+)
+
+// DigestHash generates the digest of statements.
+// it will generate a hash on normalized form of statement text
+// which removes general property of a statement but keeps specific property.
+//
+// for example: both DigestHash('select 1') and DigestHash('SelEct 2') => e1c71d1661ae46e09b7aaec1c390957f0d6260410df4e4bc71b9c8d681021471
+func DigestHash(sql string) (result string) {
+	d := digesterPool.Get().(*sqlDigester)
+	result = d.doDigest(sql)
+	digesterPool.Put(d)
+	return
+}
+
+// DigestText generates the normalized statements.
+// it will get normalized form of statement text
+// which removes general property of a statement but keeps specific property.
+//
+// for example: DigestText('select 1 from b where a = 1') => 'select ? from b where a = ?'
+func DigestText(sql string) (result string) {
+	d := digesterPool.Get().(*sqlDigester)
+	result = d.doDigestText(sql)
+	digesterPool.Put(d)
+	return
+}
+
+var digesterPool = sync.Pool{
+	New: func() interface{} {
+		return &sqlDigester{
+			lexer:  NewScanner(""),
+			hasher: sha256.New(),
+		}
+	},
+}
+
+// sqlDigester is used to compute DigestHash or DigestText for sql.
+type sqlDigester struct {
+	buffer bytes.Buffer
+	lexer  *Scanner
+	hasher hash2.Hash
+	tokens tokenDeque
+}
+
+func (d *sqlDigester) doDigest(sql string) (result string) {
+	d.normalize(sql)
+	d.hasher.Write(d.buffer.Bytes())
+	d.buffer.Reset()
+	result = fmt.Sprintf("%x", d.hasher.Sum(nil))
+	d.hasher.Reset()
+	return
+}
+
+func (d *sqlDigester) doDigestText(sql string) (result string) {
+	d.normalize(sql)
+	result = string(d.buffer.Bytes())
+	d.lexer.reset("")
+	d.buffer.Reset()
+	return
+}
+
+const (
+	// genericSymbol presents parameter holder ("?") in statement
+	// it can be any value as long as it is not repeated with other tokens.
+	genericSymbol = -1
+	// genericSymbolList presents parameter holder lists ("?, ?, ...") in statement
+	// it can be any value as long as it is not repeated with other tokens.
+
+	// it can be any value as long as it is not repeated with other tokens.
+	genericSymbolList = -2
+)
+
+func (d *sqlDigester) normalize(sql string) {
+	d.lexer.reset(sql)
+	for {
+		tok, pos, lit := d.lexer.scan()
+		if tok == unicode.ReplacementChar && d.lexer.r.eof() {
+			break
+		}
+		if pos.Offset == len(sql) {
+			break
+		}
+		if tok == hintEnd {
+			d.reduceHint()
+			continue
+		}
+		currTok := token{tok, strings.ToLower(lit)}
+		if isLit(tok) {
+			currTok = d.reduceLit(currTok)
+		}
+		d.tokens = append(d.tokens, currTok)
+	}
+	d.lexer.reset("")
+	for i, token := range d.tokens {
+		d.buffer.WriteString(token.lit)
+		if i != len(d.tokens)-1 {
+			d.buffer.WriteRune(' ')
+		}
+	}
+	d.tokens = d.tokens[:0]
+}
+
+func (d *sqlDigester) reduceHint() {
+	for {
+		token := d.tokens.popBack(1)
+		if len(token) == 0 {
+			return
+		}
+		if token[0].tok == hintBegin {
+			return
+		}
+	}
+}
```

Comment:
Why should the hint be removed during normalization?

Reply:
I'm not sure this too, maybe remove hint make SQL binding easier when we want to force plan no matter what hint in current sql?

Reply:
Is this behavior compatible with MySQL? @lamxTyler any comments on this behavior?

Reply:
/cc @morgo

Reply:
no, it's too hard compatible mysql value, but on other side, we do more things than mysql, mysql's digest doesn't handle logic like `group/order by`, because our digest is used for sql plan binding.

Reply:
ping @lamxTyler, should plan manage need that? if so I will take that :rofl:

Reply:
It would be good it if we can do that...

Reply:
> Should we also ignore the `force index()`/`use index()`/`ignore index()`/`straight_join` key words?

Then we should remove these as well?

Reply:
removed